### PR TITLE
Vanilla ICP

### PR DESCRIPTION
**Pull request by jeffin07**
#### Changes
Added Vanilla ICP 

Fixes # (issue)

#2649 
#### Type of change

- [x] 🔬 New feature (non-breaking change which adds functionality)
- [x] 📝 This change requires a documentation update

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

**Comment by jeffin07** [comment]
@edgarriba I have created a draft PR https://github.com/kornia/kornia/pull/2875. Can you please take a look at it ?

File: kornia/utils/pointcloud_io.py [hunk 1]
```diff
@@ -81,3 +81,64 @@ def load_pointcloud_ply(filename: str, header_size: int = 8) -> torch.Tensor:
         # create tensor from list
         pointcloud: torch.Tensor = torch.tensor(points)
         return pointcloud
+
+
+def iterative_closest_point(
+    points_in_a: torch.Tensor, points_in_b: torch.Tensor, max_iterations: int = 20, tolerance: float = 1e-4
```
Comment (cjpurackal) [inline kornia/utils/pointcloud_io.py:87]:
```suggestion
    points_in_a: torch.Tensor, points_in_b: torch.Tensor, max_iterations: int = 20, tolerance: float = 1e-4, verbose: bool = False
```

File: kornia/utils/pointcloud_io.py [hunk 2]
```diff
@@ -81,3 +81,64 @@ def load_pointcloud_ply(filename: str, header_size: int = 8) -> torch.Tensor:
...
+        points_in_a: The point cloud in the source coordinates frame A  with shape Nx3
+        points_in_b:  The point cloud in the source coordinates frame A with shape Nx
+        max_iterations (int): Maximum number of iterations to run.
+        tolerance (float): Tolerance criteria for stopping.
+
+     Return:
+        The relative transformation between the two pointcloud with shape 3x4
+    """
+
+    src = points_in_a.clone()
```
Comment (cjpurackal) [inline kornia/utils/pointcloud_io.py:105]: add some basic shape and type check 
`KORNIA_CHECK_SHAPE(points_in_a, ["B", "3"])`

File: kornia/utils/pointcloud_io.py [hunk 3]
```diff
@@ -81,3 +81,64 @@ def load_pointcloud_ply(filename: str, header_size: int = 8) -> torch.Tensor:
...
+
+        R = U.T @ Vt.T
+
+        if torch.det(R) < 0:
+            Vt[-1, :] *= -1
+            R = U.T @ Vt.T
+
+        t = b_mean.T - R @ a_mean.T
+
+        points_in_a = (points_in_a.T @ R + t.unsqueeze(-1)).T
```
Comment (cjpurackal) [inline kornia/utils/pointcloud_io.py:133]:
```suggestion
        points_in_a = points_in_a @ R + t
```

File: kornia/utils/pointcloud_io.py [hunk 4]
```diff
@@ -81,3 +81,64 @@ def load_pointcloud_ply(filename: str, header_size: int = 8) -> torch.Tensor:
...
+
+        mean_error = torch.mean(torch.norm(src - dst[min_idx], dim=1))
+
```
Comment (cjpurackal) [inline kornia/utils/pointcloud_io.py:136]: if verbose:
            print(f"ICP Mean error: {mean_error}")

File: kornia/utils/pointcloud_io.py [hunk 5]
```diff
@@ -81,3 +81,64 @@ def load_pointcloud_ply(filename: str, header_size: int = 8) -> torch.Tensor:
...
+        max_iterations (int): Maximum number of iterations to run.
+        tolerance (float): Tolerance criteria for stopping.
+
+     Return:
+        The relative transformation between the two pointcloud with shape 3x4
+    """
+
+    src = points_in_a.clone()
+    dst = points_in_b.clone()
+    prev_error = 0
```
Comment (cjpurackal) [inline kornia/utils/pointcloud_io.py:107]:
```suggestion
    prev_error = 0
    R = torch.eye(3)
    t = torch.zeros(3)
```

File: kornia/utils/pointcloud_io.py [hunk 6]
```diff
@@ -81,3 +81,64 @@ def load_pointcloud_ply(filename: str, header_size: int = 8) -> torch.Tensor:
...
+        mean_error = torch.mean(torch.norm(src - dst[min_idx], dim=1))
+
+        if torch.abs(prev_error - mean_error) < tolerance:
+            break
+        prev_error = mean_error
+        print(mean_error, iter)
+
+    # print(points_in_a)
+
+    return 0
```
Comment (cjpurackal) [inline kornia/utils/pointcloud_io.py:144]:
```suggestion
    return torch.cat([R, t[:, None]], dim=1)
```

File: kornia/utils/pointcloud_io.py [hunk 7]
```diff
@@ -81,3 +81,64 @@ def load_pointcloud_ply(filename: str, header_size: int = 8) -> torch.Tensor:
...
+        points_in_a = (points_in_a.T @ R + t.unsqueeze(-1)).T
+
+        mean_error = torch.mean(torch.norm(src - dst[min_idx], dim=1))
+
+        if torch.abs(prev_error - mean_error) < tolerance:
+            break
+        prev_error = mean_error
+        print(mean_error, iter)
+
+    # print(points_in_a)
```
Comment (cjpurackal) [inline kornia/utils/pointcloud_io.py:142]:
```suggestion
```

File: kornia/utils/pointcloud_io.py [hunk 8]
```diff
@@ -81,3 +81,64 @@ def load_pointcloud_ply(filename: str, header_size: int = 8) -> torch.Tensor:
...
+        t = b_mean.T - R @ a_mean.T
+
+        points_in_a = (points_in_a.T @ R + t.unsqueeze(-1)).T
+
+        mean_error = torch.mean(torch.norm(src - dst[min_idx], dim=1))
+
+        if torch.abs(prev_error - mean_error) < tolerance:
+            break
+        prev_error = mean_error
+        print(mean_error, iter)
```
Comment (cjpurackal) [inline kornia/utils/pointcloud_io.py:140]:
```suggestion
        if verbose:
            print(f"ICP Mean error: {mean_error}, iter: {iter}")
```

File: kornia/utils/pointcloud_io.py [hunk 9]
```diff
@@ -81,3 +81,64 @@ def load_pointcloud_ply(filename: str, header_size: int = 8) -> torch.Tensor:
...
+) -> torch.Tensor:
+    """Compute the relative transformation between two point clouds.
+
+    The resulting transformation uses the iterative closest point algorithm to satisfy:
+
+        points_in_b = b_from_a @ points_in_a
+
+    Args:
+        points_in_a: The point cloud in the source coordinates frame A  with shape Nx3
+        points_in_b:  The point cloud in the source coordinates frame A with shape Nx
```
Comment (cjpurackal) [inline kornia/utils/pointcloud_io.py:97]:
```suggestion
        points_in_b:  The point cloud in the source coordinates frame A with shape Nx3
```

**Comment by cjpurackal** [comment]
@jeffin07 can you write a test to verify the functionality?

**Comment by jeffin07** [comment]
@cjpurackal  Added test. Please review the changes

File: tests/utils/test_pointcloud_io.py [hunk 10]
```diff
@@ -1,5 +1,6 @@
 import os
 
+import pypose as pp
```
Comment (edgarriba) [inline tests/utils/test_pointcloud_io.py:3]: Don’t import this library, just generate the results and hardcode the resulting tensor values in the test

File: kornia/utils/pointcloud_io.py [hunk 11]
```diff
@@ -81,3 +84,66 @@ def load_pointcloud_ply(filename: str, header_size: int = 8) -> torch.Tensor:
         # create tensor from list
         pointcloud: torch.Tensor = torch.tensor(points)
         return pointcloud
+
+
+def iterative_closest_point(
```
Comment (edgarriba) [inline kornia/utils/pointcloud_io.py:89]: Can you put this into its own file under `kornia.geometry.pointcloud.icp.py` ? /cc @shijianjian @cjpurackal @johnnv1 @ducha-aiki

File: kornia/utils/pointcloud_io.py [hunk 12]
```diff
@@ -81,3 +84,66 @@ def load_pointcloud_ply(filename: str, header_size: int = 8) -> torch.Tensor:
...
+        a_center = src - a_mean
+        b_center = points_in_b[min_idx] - b_mean
+
+        H = a_center.T @ b_center
+
+        U, S, Vt = torch.linalg.svd(H)
+
+        R = Vt.T @ U.T
+
+        if torch.det(R) < 0:
```
Comment (edgarriba) [inline kornia/utils/pointcloud_io.py:134]: What if this true only for one rotation of the batch , then you apply to all ? I believe there a need of masking here

File: kornia/utils/pointcloud_io.py [hunk 13]
```diff
@@ -81,3 +84,66 @@ def load_pointcloud_ply(filename: str, header_size: int = 8) -> torch.Tensor:
...
+            Vt[-1, :] *= -1
+            R = Vt.T @ U.T
+
+        t = b_mean.T - R @ a_mean.T
+
+        src = (R @ src.T + t.unsqueeze(-1)).T
+
+        mean_error = torch.mean(torch.norm(src - points_in_b[min_idx], dim=1))
```
Comment (edgarriba) [inline kornia/utils/pointcloud_io.py:142]: This is the error over the batch, right ? Same I think masking is needed, or this algorithm just runs for a single set of points, then we batch with loop or vmap @cjpurackal  what do you think ?

**Review by edgarriba — CHANGES_REQUESTED** [review]
Not sure the rebase was done properly, the diff shows other recent changes.

**Review by edgarriba — COMMENTED** [review]
For later, but a visual tutorial would be great.

File: kornia/utils/pointcloud_io.py (hunk 11 above)
Comment (johnnv1) [reply kornia/utils/pointcloud_io.py:89]: +1

Comment (jeffin07) [reply kornia/utils/pointcloud_io.py]: Done

File: kornia/utils/pointcloud_io.py (hunk 12 above)
Comment (jeffin07) [reply kornia/utils/pointcloud_io.py]: @edgarriba  Currently we only consider a set of points. Sorry for the delay

File: kornia/utils/pointcloud_io.py (hunk 13 above)
Comment (jeffin07) [reply kornia/utils/pointcloud_io.py]: @edgarriba  Yes, Currently it runs for a single set of points
